clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1096,96 +1097,156 @@ public void assign(Collection<TopicPartition> partitions) {
                 }
+                final long fetchEnd = time.milliseconds();
+                remainingMs = Math.max(0L, remainingMs - (fetchEnd - fetchStart));
 
-                long elapsed = time.milliseconds() - start;
-                remaining = timeout - elapsed;
-            } while (remaining > 0);
+            } while (remainingMs > 0);
 
             return ConsumerRecords.empty();
         } finally {
             release();
         }
     }
 
+
     /**
-     * Do one round of polling. In addition to checking for new data, this does any needed offset commits
-     * (if auto-commit is enabled), and offset resets (if an offset reset policy is defined).
-     * @param timeout The maximum time to block in the underlying call to {@link ConsumerNetworkClient#poll(long)}.
-     * @return The fetched records (may be empty)
+     * Visible for testing
      */
-    private Map<TopicPartition, List<ConsumerRecord<K, V>>> pollOnce(long timeout) {
-        client.maybeTriggerWakeup();
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted") // because false => timed out, in which case we return early or throw.
+    boolean updateAssignmentMetadataIfNeeded(final Duration timeout) {
+        final long startMs = time.milliseconds();
+        if (!coordinator.poll(remainingMsAtLeastZero(startMs, timeout))) {
+            return false;
+        }
 
-        long startMs = time.milliseconds();
-        coordinator.poll(startMs, timeout);
+        return updateFetchPositions(remainingMsAtLeastZero(startMs, timeout));
+    }
 
-        // Lookup positions of assigned partitions
-        boolean hasAllFetchPositions = updateFetchPositions();
+    private Map<TopicPartition, List<ConsumerRecord<K, V>>> pollForFetches(final long upperBoundTimeoutMs) {
+        final long startMs = time.milliseconds();
+        long pollTimeout = Math.min(coordinator.timeToNextPoll(startMs), upperBoundTimeoutMs);
 
         // if data is available already, return it immediately
-        Map<TopicPartition, List<ConsumerRecord<K, V>>> records = fetcher.fetchedRecords();
-        if (!records.isEmpty())
+        final Map<TopicPartition, List<ConsumerRecord<K, V>>> records = fetcher.fetchedRecords();
+        if (!records.isEmpty()) {
             return records;
+        }
 
         // send any new fetches (won't resend pending fetches)
         fetcher.sendFetches();
 
-        long nowMs = time.milliseconds();
-        long remainingTimeMs = Math.max(0, timeout - (nowMs - startMs));
-        long pollTimeout = Math.min(coordinator.timeToNextPoll(nowMs), remainingTimeMs);
-
         // We do not want to be stuck blocking in poll if we are missing some positions
         // since the offset lookup may be backing off after a failure
-        if (!hasAllFetchPositions && pollTimeout > retryBackoffMs)
+        if (!subscriptions.hasAllFetchPositions() && pollTimeout > retryBackoffMs) {
```

Comment:
hmm. I'm not sure what a good way to do this is... `hasAllFetchPositions` was the result of `updateAllFetchPositions` before (which changed because the result now indicates whether it timed out or not). But `updateAllFetchPositions` is not longer called in this scope, so we can't consider keeping its result around.

I could cache the "hasAllFetchPositions" fact on a field in KafkaConsumer, but it seems a little risky. I *think* it would be fine... WDYT?